src/requests/features.jl

```diff
@@ -114,15 +114,116 @@ function get_file_loc(x::EXPR, offset=0, c=nothing)
     end
 end
 
+function search_file(filename, dir, topdir)
+    parent_dir = dirname(dir)
+    return if (!startswith(dir, topdir) || parent_dir == dir || isempty(dir))
+        nothing
+    else
+        path = joinpath(dir, filename)
+        isfile(path) ? path : search_file(filename, parent_dir, topdir)
+    end
+end
+
+function get_juliaformatter_config(doc, server)
+    path = get_path(doc)
+
+    # search through workspace for a `.JuliaFormatter.toml`
+    workspace_dirs = sort(filter(f -> startswith(path, f), collect(server.workspaceFolders)), by = length, rev = true)
+    config_path = length(workspace_dirs) > 0 ?
+        search_file(JuliaFormatter.CONFIG_FILE_NAME, path, workspace_dirs[1]) :
+        nothing
+
+    config_path === nothing && return nothing
+
+    @debug "Found JuliaFormatter config at $(config_path)"
+    return JuliaFormatter.parse_config(config_path)
+end
+
 function textDocument_formatting_request(params::DocumentFormattingParams, server::LanguageServerInstance, conn)
     doc = getdocument(server, URI2(params.textDocument.uri))
-    newcontent = DocumentFormat.format(get_text(doc), server.format_options)
+
+    config = get_juliaformatter_config(doc, server)
+
+    newcontent = if config === nothing
+        JuliaFormatter.format_text(get_text(doc); indent=params.options.tabSize)
+    else
+        JuliaFormatter.format_text(get_text(doc); JuliaFormatter.kwargs(config)...)
+    end
+
     end_l, end_c = get_position_at(doc, sizeof(get_text(doc))) # AUDIT: OK
     lsedits = TextEdit[TextEdit(Range(0, 0, end_l, end_c), newcontent)]
 
     return lsedits
 end
 
+function textDocument_range_formatting_request(params::DocumentRangeFormattingParams, server::LanguageServerInstance, conn)
+    doc = getdocument(server, URI2(params.textDocument.uri))
+    cst = getcst(doc)
+
+    expr = get_inner_expr(cst, get_offset(doc, params.range.start):get_offset(doc, params.range.stop))
+
+    if expr === nothing
+        return nothing
+    end
+
+    while !(expr.head in (:for, :if, :function, :module, :file))
+        if expr.parent !== nothing
+            expr = expr.parent
+        else
+            return nothing
+        end
+    end
+
+    _, offset = get_file_loc(expr)
+    l1, c1 = get_position_at(doc, offset)
+    c1 = 0
+    start_offset = get_offset2(doc, l1, c1)
+    l2, c2 = get_position_at(doc, offset + expr.span)
+    end_offset = get_offset(doc, l2, c2)
+
+    text = get_text(doc)[start_offset:end_offset]
+
+    longest_prefix = nothing
+    for line in eachline(IOBuffer(text))
+        (isempty(line) || occursin(r"^\s*$", line)) && continue
+        idx = 0
+        for c in line
+            if c == ' ' || c == '\t'
+                idx += 1
+            else
+                break
+            end
+        end
+        line = line[1:idx]
+        longest_prefix = CSTParser.longest_common_prefix(something(longest_prefix, line), line)
+    end
+
+    config = get_juliaformatter_config(doc, server)
```

Comment:
Does this request get hit every key stroke? If so it's not ideal to search the disk each time. I can't remember whether we're able to have a file watch from within the languageserver which may be why this has been done

Reply:
We could just add those files to https://github.com/julia-vscode/julia-vscode/blob/master/src/extension.ts#L232, and then we would get normal LSP messages about their content etc. That seems in general better?

Reply:
> Does this request get hit every key stroke?

No, only when explicitly requested.

> We could just add those files to https://github.com/julia-vscode/julia-vscode/blob/master/src/extension.ts#L232, and then we would get normal LSP messages about their content etc. That seems in general better?

IIUC we'd only be notified by FS events or when the user opens/changes the `.JuliaFormatter.toml`. The client won't just dump all matching files into the server.